File: haskell/complex-numbers/src/ComplexNumbers.hs
module ComplexNumbers (
  Complex,
  conjugate,
  abs,
  exp,
  real,
  imaginary,
  mul,
  add,
  sub,
  div,
  complex,
) where

import Prelude hiding (abs, div, exp)
import qualified Prelude as P

-- Data definition -------------------------------------------------------------
data Complex a = Complex a a deriving (Show, Eq)

complex :: (a, a) -> Complex a
complex (a, b) = Complex a b

-- unary operators -------------------------------------------------------------
conjugate :: Num a => Complex a -> Complex a
conjugate (Complex a b) = Complex a (-b)

abs :: Floating a => Complex a -> a
abs (Complex a b) = sqrt $ a ^ 2 + b ^ 2

real :: Num a => Complex a -> a
real (Complex a _) = a 

imaginary :: Num a => Complex a -> a
imaginary (Complex _ b) = b

exp :: Floating a => Complex a -> Complex a
exp (Complex a b) = Complex (P.exp a * cos b) (P.exp a * sin b)

-- binary operators ------------------------------------------------------------
mul :: Num a => Complex a -> Complex a -> Complex a
mul (Complex a b) (Complex a' b') = Complex (a * a' - b * b') (a * b' + b * a') 

add :: Num a => Complex a -> Complex a -> Complex a
add (Complex a b) (Complex a' b') = Complex (a + a') (b + b')

sub :: Num a => Complex a -> Complex a -> Complex a
sub (Complex a b) (Complex a' b') = Complex (a - a') (b - b')

div :: Fractional a => Complex a -> Complex a -> Complex a
div (Complex a b) (Complex a' b') = Complex ((a * a' + b * b') / (a' ^ 2 + b' ^ 2)) ((b * a' - a * b') / (a' ^ 2 + b' ^ 2))

File: typescript/complex-numbers/complex-numbers.ts
export class ComplexNumber {
  readonly _real: number;
  readonly _imaginary: number;

  constructor(real: number, imaginary: number) {
    this._real = real;
    this._imaginary = imaginary;
  }

  public get real(): number {
    return this._real;
  }

  public get imag(): number {
    return this._imaginary;
  }

  public add(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(
      this._real + other.real,
      this._imaginary + other.imag,
    );
  }

  public sub(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(
      this._real - other.real,
      this._imaginary - other.imag,
    );
  }

  public div(other: ComplexNumber): ComplexNumber {
    const denominator = other.real * other.real + other.imag * other.imag;
    if (denominator === 0) {
      throw new Error("Division by zero");
    }

    const realPart =
      (this._real * other.real + this._imaginary * other.imag) / denominator;
    const imagPart =
      (this._imaginary * other.real - this._real * other.imag) / denominator;

    return new ComplexNumber(realPart, imagPart);
  }

  public mul(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(
      this._real * other.real - this._imaginary * other.imag,
      this._real * other.imag + this._imaginary * other.real,
    );
  }

  public get abs(): number {
    return Math.sqrt(
      this._real * this._real + this._imaginary * this._imaginary,
    );
  }

  public get conj(): ComplexNumber {
    return new ComplexNumber(
      this._real,
      this._imaginary === 0 ? 0 : -1 * this._imaginary,
    );
  }

  public get exp(): ComplexNumber {
    const expReal = Math.exp(this._real);
    return new ComplexNumber(
      expReal * Math.cos(this._imaginary),
      expReal * Math.sin(this._imaginary),
    );
  }
}
